File: cpp/508.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int helper(TreeNode *t, unordered_map<int, int> &m) {
    if (!t)
      return 0;
    int l = helper(t->left, m), r = helper(t->right, m);
    int sum = t->val + l + r;
    m[sum]++;
    return sum;
  }
  vector<int> findFrequentTreeSum(TreeNode *root) {
    unordered_map<int, int> freq;
    helper(root, freq);
    int maxFreq = 0;
    vector<int> ret;
    for (auto &t : freq) {
      if (t.second == maxFreq) {
        ret.push_back(t.first);
      } else if (t.second > maxFreq) {
        ret.clear();
        maxFreq = t.second;
        ret.push_back(t.first);
      }
    }
    return ret;
  }
};

File: go/508.go
package main

//Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func helper(t *TreeNode, m map[int]int, l *int) int {
  if t == nil {
    return 0
  }
  var sum int = t.Val + helper(t.Left, m, l) + helper(t.Right, m, l)
  m[sum]++
  if m[sum] > *l {
    *l = m[sum]
  }
  return sum
}

func findFrequentTreeSum(root *TreeNode) []int {
  var ret []int = make([]int, 0)
  var m map[int]int = make(map[int]int)
  var largest int = 0
  helper(root, m, &largest)
  for k, v := range m {
    if v == largest {
      ret = append(ret, k)
    }
  }
  return ret
}
